src/librustc_typeck/check/compare_method.rs

```diff
@@ -0,0 +1,283 @@
+    // trait_to_skol_substs and use it to convert the predicates contained
+    // in the trait_m.generics to the skolemized form.
+    //
+    // Finally we register each of these predicates as an obligation in
+    // a fresh FulfillmentCtxt, and invoke select_all_or_error.
+    //
+    // FIXME(jroesch): Currently the error reporting is not correctly attached
+    // to a span and results in terrible error reporting.
+
+    // Create a parameter environment that represents the implementation's
+    // method.
+    let impl_env = ty::ParameterEnvironment::for_item(tcx, impl_m.def_id.node);
+
+    // Create mapping from impl to skolemized.
+    let impl_to_skol_substs = &impl_env.free_substs;
+
+    // Create mapping from trait to skolemized.
+    let trait_to_skol_substs =
+        trait_to_impl_substs
+        .subst(tcx, impl_to_skol_substs)
+        .with_method(impl_to_skol_substs.types.get_slice(subst::FnSpace).to_vec(),
+                     impl_to_skol_substs.regions().get_slice(subst::FnSpace).to_vec());
+    debug!("compare_impl_method: trait_to_skol_substs={}",
+           trait_to_skol_substs.repr(tcx));
+
+    // Create a fresh fulfillment context.
+    let mut fulfill_cx = traits::FulfillmentContext::new();
+
+    // Create obligations for each predicate declared by the impl
+    // definition in the context of the trait's parameter
+    // environment. We can't just use `impl_env.caller_bounds`,
+    // however, because we want to replace all late-bound regions with
+    // region variables.
+    let impl_bounds =
+        impl_m.generics.to_bounds(tcx, impl_to_skol_substs);
+    let (impl_bounds, _) =
+        infcx.replace_late_bound_regions_with_fresh_var(
+            impl_m_span,
+            infer::HigherRankedType,
+            &ty::Binder(impl_bounds)); // TODO I think this is right?
+    debug!("compare_impl_method: impl_bounds={}",
+           impl_bounds.repr(tcx));
+
+    let trait_bounds = trait_m.generics.to_bounds(tcx, &trait_to_skol_substs);
+
+    let trait_pred = trait_bounds.predicates.split();
+    let impl_pred = impl_bounds.clone().predicates.split();
+
+    let hybrid_preds = VecPerParamSpace::new(
+        impl_pred.types,
+        impl_pred.selfs,
+        trait_pred.fns
+    );
+
+    // Construct trait parameter environment and then shift it into the skolemized viewpoint.
+    let mut trait_param_env =
+        ty::construct_parameter_environment(tcx,
+            &trait_m.generics,
+            impl_m_body_id);
+
+    let mut trait_param_env =
+        trait_param_env.subst(tcx, &trait_to_skol_substs);
+        debug!("compare_impl_method: trait_param_env={}",
+            trait_param_env.repr(tcx));
+
+    trait_param_env.caller_bounds.predicates = hybrid_preds;
+
+    debug!("compare_impl_method: trait_bounds={}",
+        trait_param_env.caller_bounds.repr(tcx));
+        
+    for predicate in impl_bounds.predicates.into_iter() {
```

Comment:
I'd say just iterate over `impl_pred.fns.into_iter()`, which conveniently (and not coincidentally) have not been moved yet :)